posthog/temporal/workflows/squash_person_overrides.py

```diff
@@ -0,0 +1,698 @@
+        return self._latest_created_at
+
+    @latest_created_at.setter
+    def latest_created_at(self, v: datetime | str | None):
+        if isinstance(v, datetime):
+            self._latest_created_at = v.isoformat()
+        else:
+            self._latest_created_at = v
+
+    def iter_person_overides_to_delete(self) -> Iterable[SerializablePersonOverrideToDelete]:
+        """Iterate over SerializablePersonOverrideToDelete ensuring they are of that type.
+
+        Looking at the types, this seems pointless, just iterate over person_overrides_to_delete!
+        However, as Temporal passes inputs to and from activities, namedtuples will be cast to
+        lists. This method thus exists to transform them back into namedtuples.
+        """
+        for person_override_to_delete in self.person_overrides_to_delete:
+            yield SerializablePersonOverrideToDelete(*person_override_to_delete)
+
+
+@activity.defn
+async def prepare_dictionary(inputs: QueryInputs) -> str:
+    """Prepare the DICTIONARY to be used in the squash workflow.
+
+    We also lock in the latest merged_at to ensure we do not process overrides that arrive after
+    we have started the job.
+    """
+    activity.logger.info("Preparing DICTIONARY %s.%s", inputs.database, inputs.dictionary_name)
+
+    latest_created_at = sync_execute(SELECT_LATEST_CREATED_AT_QUERY.format(database=inputs.database))[0][0]
+    activity.logger.info("Populating underlying TABLE %s.%s_table", inputs.database, inputs.dictionary_name)
+    sync_execute(
+        CREATE_DICTIONARY_TABLE_QUERY.format(
+            database=inputs.database,
+            dictionary_name=inputs.dictionary_name,
+            user=inputs.user,
+            password=inputs.password,
+            cluster_name=inputs.cluster_name,
+        )
+    )
+
+    sync_execute(
+        INSERT_INTO_DICTIONARY_TABLE_QUERY.format(
+            database=inputs.database,
+            dictionary_name=inputs.dictionary_name,
+            user=inputs.user,
+            password=inputs.password,
+            cluster_name=inputs.cluster_name,
+            team_id_filter="AND team_id in %(team_ids)s" if inputs.team_ids else "",
+        ),
+        {"latest_created_at": latest_created_at, "team_ids": inputs.team_ids},
+    )
+
+    activity.logger.info("Creating DICTIONARY %s.%s", inputs.database, inputs.dictionary_name)
+    sync_execute(
+        CREATE_DICTIONARY_QUERY.format(
+            database=inputs.database,
+            dictionary_name=inputs.dictionary_name,
+            user=inputs.user,
+            password=inputs.password,
+            cluster_name=inputs.cluster_name,
+        )
+    )
+
+    return latest_created_at.isoformat()
+
+
+@activity.defn
+async def select_persons_to_delete(inputs: QueryInputs) -> list[SerializablePersonOverrideToDelete]:
+    """Select the persons we'll override to lock them in and safely delete afterwards
+
+    New overrides may come in while we are executing this workflow, so we need to
```

Comment:
ah ok nice you do mention this here. however, best is to just stop them coming in I think, also because we'll then mutate the table. although the created_at guards are still useful to have.

Reply:
I kept the guards even though we are now simply detaching the table.